tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_hemv_kk.hpp

```diff
@@ -0,0 +1,290 @@
+
+#ifndef LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_HERMITIAN_MATVEC_HPP_
+#define LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_HERMITIAN_MATVEC_HPP_
+
+namespace KokkosKernelsSTD {
+
+namespace hermitianmatvecimpl{
+
+template<class T> struct is_complex : std::false_type{};
+template<> struct is_complex<std::complex<float>> : std::true_type{};
+template<> struct is_complex<std::complex<double>> : std::true_type{};
+template<> struct is_complex<std::complex<long double>> : std::true_type{};
+
+template<class T> inline constexpr bool is_complex_v = is_complex<T>::value;
+
+} // end namespace hermitianmatvecimpl
+
+//
+// Overwriting hermitian matrix-vector product: y = Ax
+//
+// for now, specialize for default_accessor
+// https://github.com/kokkos/stdBLAS/issues/122
+//
+template<class ExeSpace,
+	 class ElementType_A,
+         std::experimental::extents<>::size_type numRows_A,
+         std::experimental::extents<>::size_type numCols_A,
+         class Layout_A,
+         class Triangle,
+         class ElementType_x,
+         std::experimental::extents<>::size_type ext_x,
+         class Layout_x,
+         class ElementType_y,
+         std::experimental::extents<>::size_type ext_y,
+         class Layout_y>
+requires (Layout_A::template mapping<std::experimental::extents<numRows_A, numCols_A>>::is_always_unique())
+void hermitian_matrix_vector_product(kokkos_exec<ExeSpace> kexe,
+				      std::experimental::mdspan<
+					ElementType_A,
+					std::experimental::extents<numRows_A, numCols_A>,
+					Layout_A,
+				        std::experimental::default_accessor<ElementType_A>
+				      > A,
+				      Triangle tr,
+				      std::experimental::mdspan<ElementType_x,
+					std::experimental::extents<ext_x>,
+					Layout_x,
+					std::experimental::default_accessor<ElementType_x>
+				      > x,
+				      std::experimental::mdspan<ElementType_y,
+					std::experimental::extents<ext_y>,
+					Layout_y,
+					std::experimental::default_accessor<ElementType_y>
+				      > y)
+{
+
+  // constraints
```

Comment:
Please see my comments elsewhere on the difference between Constraints and Mandates; thanks!

Reply:
I realize also that these `static_assert`s will never fail, because this overload only participates in overload resolution if they are all true.  Thus, I think I see what you're trying to do, which is double-check your constraints (rather than impose new Mandates conditions).  That's not bad, but it's unnecessary; the properties already hold by construction.

Reply:
Would it make sense to follow your suggestion by replacing these `static_assert`s with something like:
```cpp
template<...>
  // P1673 constraints
  requires (requires () { decltype(A)::rank() == 2; })
        && (requires () { decltype(x)::rank() == 1; })
        && (requires () { decltype(y)::rank() == 1; })
void hermitian_matrix_vector_product(...)
```
?

It duplicates what `mdspan`/`extents` do in the function header now, but:
1. gives clear reference to and description of specification _Constraints_;
2. will alarm if someone unconsciously relaxes the constraints by changing the header (e.g. generalizing the types).

Alternatively, I thought of a comment like 
```cpp
// Note: P1673 constraints (rank 2 for A and rank 1 for x and y)
// are specified with mdspan<..., extents<*>, ...> in the function header
template<...>
void hermitian_matrix_vector_product(...)
```
but `requires` seems much better.

> PS. You touched on asserts duplicating the overload spec, so I'm asking here in context of https://github.com/kokkos/stdBLAS/pull/157#discussion_r846881993

Reply:
@MikolajZuzek Thanks for the suggestions! : - )  I would actually prefer nothing at all -- neither option -- based on the Don't Repeat Yourself (DRY) Principle.  The function's parameters already express the desired constraints.  If those ever need to change, then you would only have one place to change them.

If I had to pick one of the two options, it would be the `static_assert`s.  Adding redundant `requires` clauses would make debugging harder.  (It would actually shift debugging to run time, because you'll need the logging output to see that you're not getting the kokkos-kernels overload.)

Btw, the `requires` clauses can be spelled more concisely as (e.g.,) `requires(A.rank() == 2)`.  [Here](https://godbolt.org/z/aWj8e7Yvs) is an example.
```c++
#include <array>

template<class T>
T f(T t) requires(t.size() == 3)
{
    return t;
}

int main() {
    std::array<int, 3> x;
    std::array<int, 4> y;
    std::array<int, 5> z;

    auto x2 = f(x);
    //auto y2 = f(y); // no matching function
    //auto z2 = f(z); // no matching function

    return 0;
}
```